sqlglot/parser.py

```diff
@@ -1867,7 +1870,10 @@ def _parse_query_modifiers(self, this: t.Optional[exp.Expression]) -> None:
             join = self._parse_join()
             comma = None if table else self._match(TokenType.COMMA)
             if lateral:
-                this.append("laterals", lateral)
+                if isinstance(lateral, exp.Join):
```

Comment:
why was this changed, along with the optimizer?

Reply:
This was changed to keep the correct order of joins. When adding it to the laterals, the lateral joins are executed after the joins.

Reply:
As for the optimizer. The laterals coming from lateral joins werent traversed correctly.

Reply:
why is a lateral a join?

Reply:
Well when using a lateral join it is a join in my opinion. Lateral view however is not a join. 

Im using it in spark:
https://spark.apache.org/docs/latest/sql-ref-syntax-qry-select-join.html

Reply:
so if this is a lateral join, shouldn't this be a lateral? how is a lateral join a join? if that's true maybe we should fix the join parser to parse laterals, instead of fixing it here

Reply:
Well as i see it, its a join on a lateral subquery or table valued function. For example im able to do LEFT JOIN LATERAL or a JOIN LATERAL in spark. So i would see it as a nested reference in those cases.